src/lib/e2esdk.ts

```diff
@@ -0,0 +1,204 @@
+  decryptFileContents,
+  FileMetadata,
+  Sodium,
+} from "@socialgouv/e2esdk-crypto";
+
+import { Client } from "@socialgouv/e2esdk-client";
+
+import { decryptedDataSchema } from "../../pages/form";
+import { fetchHasura } from "./hasura";
+import { insert_one } from "../queries/form";
+
+const formMetadata = z.object({
+  id: z.number(),
+  created_at: z.string(),
+  public_key: z.string(),
+  sealed_secret: z.string(),
+  signature: z.string(),
+});
+
+export type EncryptedAnswer = z.infer<typeof formMetadata> & { data: string };
+
+type SubmissionQueryVariables = {
+  submissionBucketId: string;
+  signature: string;
+  sealedSecret: string;
+  publicKey: string;
+  answersFiles: any[];
+  data: string;
+};
+
+/**
+ * Download a file based on its metadata :
+ * - use the hash to download encrypted blob
+ * - use the metadata to decipher
+ * todo: extract to e2esdk ?
+ */
+export async function downloadAndDecryptFile(
+  sodium: Sodium,
+  metadata: FileMetadata
+) {
+  const res = await fetch(`/api/storage?hash=${metadata.hash}`);
+  const blob = await res.blob();
+  const cleartext = decryptFileContents(
+    sodium,
+    new Uint8Array(await blob.arrayBuffer()),
+    {
+      algorithm: "secretBox",
+      key: sodium.from_base64(metadata.key),
+    }
+  );
+  return new File([cleartext], metadata.name, {
+    type: metadata.type,
+    lastModified: metadata.lastModified,
+  });
+}
+
+/**
+ * Decrypt a given answer
+ * - decrypt the encrypted values
+ * - parse and validate with zod
+ * todo: extract to e2esdk
+ */
+export const decryptAnswer = (
+  client: Client,
+  nameFingerprint: string,
+  answer: EncryptedAnswer
+) => {
+  try {
+    const values = client.unsealFormData(
+      {
+        metadata: {
+          publicKey: answer.public_key,
+          sealedSecret: answer.sealed_secret,
+          signature: answer.signature,
+        },
+        encrypted: { data: answer.data },
+      },
+      nameFingerprint
+    ) as Record<"data", string>;
+
+    const decryptedValues: FormData = JSON.parse(values.data);
+    const res = decryptedDataSchema.safeParse(decryptedValues);
+    if (!res.success) {
+      console.error(`Zod: Impossible de parser la réponse ${answer.id}`);
+      console.error(res.error);
+      // warning : returning null here is recommended to avoid security issues where malicious content is sent that could break the rendering process, and lead to a blank page.
+      return null;
+    }
+    return {
+      ...answer,
+      ...res.data,
+      data: undefined,
+    };
+  } catch (e) {
+    console.error(`e2esdk: Impossible de parser la réponse ${answer.id}`);
+    console.error(e);
+    return null;
+  }
+};
+
+/**
+ * return encrypted version and metadata for a given `File`
+ * todo: extract to e2esdk
+ */
+export const readAndEncryptFile = async (
+  file: File,
+  encryptionState: EncryptedFormLocalState
+): Promise<{ encryptedFile: File; metadata: Record<string, any> }> =>
```

Comment:
**suggestion**: The following would preserve typing of the output as defined in the `@socialgouv/e2esdk-crypto` package:
```suggestion
): ReturnType<typeof encryptFile> =>
```

Reply:
` Promise<ReturnType<typeof encryptFile>>` ?